### PR TITLE
Mitaka nfp 4 july 2016 opensource gerrit sync asav changes

### PR DESCRIPTION
**Pull request by sdharav**
@njagadish @mak-454 please review the changes

### CODE REVIEW

**Comment by chenchenWang** [comment]
If no rule is in template, then paloalto fw chain is able to launch
successfully.
But when there is rule in template,  due to some non-synchronized thing,
create_firewall is called before configure_interface, and create_firewall
fails because at this moment fw is not ready for configuration commands.

Another thing is, as the chain is launched. Isn't fw should sitting between
router and provider instance? Currently, on network topology, fw has only
two interfaces, and does not have an stitching interface on router,
although fw attributes shows there are three interfaces.

[image: Inline image 2]

[image: Inline image 1]

**Comment by AkashDeepSrivastava** [comment]
1) If there are no rule, configure firewall wont be called in configurator.so ,chain will be created successfully.
2) create_firewall is called before configure_interface is already fixed,but it is only for service_vendor "asav".can you please tell, what is the service vendor for paloalto fw chain.

**Comment by chenchenWang** [comment]
paloalto_fw_profile:

vendor: NFP

service flavor: service_vendor=paloalto,device_type=nova

File: gbpservice/nfp/configurator/drivers/firewall/asav/asav_fw_driver.py [hunk 1]
```diff
@@ -276,6 +312,14 @@ def configure_interfaces(self, context, resource_data):
             LOG.error(msg)
             raise Exception(msg)
 
+        try:
+            status = self._log_forwarding(mgmt_ip)
```
Comment (njagadish) [inline gbpservice/nfp/configurator/drivers/firewall/asav/asav_fw_driver.py]: We need to check unsuccessful error status and handle error just like what is done for exception.

File: gbpservice/nfp/configurator/drivers/firewall/asav/asav_fw_driver.py [hunk 2]
```diff
@@ -237,6 +237,42 @@ def _get_device_interface_name(self, cidr):
...
+            return False
+
+        url = const.REQUEST_URL % (mgmt_ip_address, "/api/cli")
+
+        enable_trap_data = {
+            "commands": [
+                "logging trap informational"
+            ]
+        }
+        status = self.rest_api.post(url, enable_trap_data, self.auth)
```
Comment (njagadish) [inline gbpservice/nfp/configurator/drivers/firewall/asav/asav_fw_driver.py]: If the return value of this function is True/False, then check the status and return accordingly.

File: gbpservice/nfp/configurator/drivers/firewall/asav/asav_fw_driver.py [hunk 3]
```diff
@@ -276,6 +312,14 @@ def configure_interfaces(self, context, resource_data):
             LOG.error(msg)
             raise Exception(msg)
 
+        try:
+            status = self._log_forwarding(mgmt_ip)
+        except Exception:
+            msg = ("Configuration of log forwarding has failed. Reason: %r" %
+                   status.json())
+            LOG.error(msg)
+            return status.json()
```
Comment (njagadish) [inline gbpservice/nfp/configurator/drivers/firewall/asav/asav_fw_driver.py]: Lets not fail the entire interface configuration because of failure in log forwarding setup. Log error and continue.

File: gbpservice/nfp/orchestrator/drivers/orchestration_driver.py [hunk 4]
```diff
@@ -634,6 +640,10 @@ def create_network_function_device(self, device_data,
                                           by=len(interfaces))
             return None
 
+        vendor_data = vendor_data_result.get('result', None)
```
Comment (njagadish) [inline gbpservice/nfp/orchestrator/drivers/orchestration_driver.py:642]: Is this result set to None when there is an exception in the function that is registered with job executor?

Comment (mak-454) [inline gbpservice/nfp/orchestrator/drivers/orchestration_driver.py:642]: no it will empty dict

File: gbpservice/nfp/configurator/drivers/firewall/asav/asav_fw_driver.py (hunk 3 above)
Comment (mak-454) [inline gbpservice/nfp/configurator/drivers/firewall/asav/asav_fw_driver.py]: Please return success here as it is done for haproxy.. sometimes we observed log forwarding failing for haproxy.

File: gbpservice/nfp/orchestrator/drivers/orchestration_driver.py (hunk 4 above)
Comment (mak-454) [inline gbpservice/nfp/orchestrator/drivers/orchestration_driver.py:642]: Jagadish, no, it will be set to empty dictionary. So, no issue here as we are doing 'get' on empty dictionary.

File: gbpservice/nfp/orchestrator/drivers/orchestration_driver.py [hunk 5]
```diff
@@ -326,6 +326,7 @@ def _update_vendor_data(self, device_data, token=None):
                     vendor_data,
                     nfp_constants.SUPPORTS_HOTPLUG)
             else:
+
```
Comment (mak-454) [inline gbpservice/nfp/orchestrator/drivers/orchestration_driver.py]: remove this blank line.

File: gbpservice/nfp/orchestrator/drivers/orchestration_driver.py [hunk 6]
```diff
@@ -634,6 +640,10 @@ def create_network_function_device(self, device_data,
                                           by=len(interfaces))
             return None
 
+        vendor_data = vendor_data_result.get('result', None)
+        if not vendor_data:
+            LOG.exception(_LE('Failed to get vendor data for device creation.'))
+
```
Comment (mak-454) [inline gbpservice/nfp/orchestrator/drivers/orchestration_driver.py:645]: should this be exception ? I think we discussed this last time, vyos will not have vendor data right.. I think it is ok not to have vendor data, so debug/warn should be fine i suppose ?

File: gbpservice/nfp/orchestrator/modules/device_orchestrator.py [hunk 7]
```diff
@@ -554,6 +554,19 @@ def _update_device_data(self, device, device_data):
         device.update(device_data)
         return device
 
+    def _make_ports_dict(self, consumer, provider, port_type):
+
+        t_ports = []
+        for ptg in [consumer, provider]:
+            if ptg[port_type]:
```
Comment (mak-454) [inline gbpservice/nfp/orchestrator/modules/device_orchestrator.py]: not the right way to check.. instead check 'if port_type in ptg.keys()'

File: gbpservice/nfp/orchestrator/modules/device_orchestrator.py [hunk 8]
```diff
@@ -983,25 +990,17 @@ def plug_interfaces_fast(self, event):
...
-            ports.append(
-                {'id': consumer['port']['id'],
-                 'port_classification': consumer['port_classification'],
-                 'port_model': consumer['port_model']})
-        if provider['port']:
-            ports.append(
-                {'id': provider['port']['id'],
-                 'port_classification': provider['port_classification'],
-                 'port_model': provider['port_model']})
+        ports = self._make_ports_dict(consumer, provider, 'port')
```
Comment (mak-454) [inline gbpservice/nfp/orchestrator/modules/device_orchestrator.py:989]: why call again here ? self._make_ports_dict is already called above, 'ports' are assigned above.

File: gbpservice/nfp/configurator/drivers/firewall/asav/asav_fw_driver.py (hunk 3 above)
Comment (ashutosh-mishra) [inline gbpservice/nfp/configurator/drivers/firewall/asav/asav_fw_driver.py]: @njagadish   - Can we add retires here ?
For enterprise mode, where visibility is expecting log forwarding to be enable, let say due to some error, log forwarding fails, still we are proceeding and launching the chain.
Is this a valid case ? Is log forwarding a necessary condition for visibility, If yes then we can't proceed as asav vendor will be present only in enterprise mode.

Comment (njagadish) [inline gbpservice/nfp/configurator/drivers/firewall/asav/asav_fw_driver.py]: Even after adding retries, there can be a case where this log forwarding can fail because of a persistent issue. So, if we fail the chain for log forwarding issue, it doesn't look good from the perspective of chain creation. From the perspectivie of visibility, it is a serious issue but, that is a bug we need to fix (just like VIP create failure in case of normal chain create).

One thing, we need to check is, why this API is failing. Once we know the nature of failure, we can take a decision on whether retries help or not.

File: gbpservice/nfp/orchestrator/drivers/orchestration_driver.py (hunk 4 above)
Comment (ashutosh-mishra) [inline gbpservice/nfp/orchestrator/drivers/orchestration_driver.py:642]: What does vendor_data_results contains ? From where 'result' key is getting added in vendor_data_result ?
If key is always present, then why do we need to do  vendor_data_result.get('result', None), we can perform vendor_data_result['result'], 
Also by default get returns None, no need to explicitly specify None.

File: gbpservice/nfp/orchestrator/drivers/orchestration_driver.py [hunk 9]
```diff
@@ -995,24 +1021,24 @@ def plug_network_function_device_interfaces(self, device_data,
                             port['id'])
                         break
 
-            for port in device_data['ports']:
-                if port['port_classification'] == nfp_constants.CONSUMER:
-                    service_type = device_data[
-                        'service_details']['service_type'].lower()
-                    if service_type.lower() in [
```
Comment (ashutosh-mishra) [inline gbpservice/nfp/orchestrator/drivers/orchestration_driver.py:1002]: @SureshDH - Didn't got this change, why we are doing this ?
This is a hotplug support case

File: gbpservice/nfp/orchestrator/modules/device_orchestrator.py [hunk 10]
```diff
@@ -578,17 +591,10 @@ def _prepare_device_data_from_nfp_context(self, nfp_context):
...
-                'id': consumer['port']['id'],
-                'port_classification': consumer['port_classification'],
-                'port_model': consumer['port_model']})
-
-        if provider['port']:
-            ports.append({
-                'id': provider['port']['id'],
-                'port_classification': provider['port_classification'],
-                'port_model': provider['port_model']})
+        if consumer['port_model'] == 'gbp_policy_target':
```
Comment (ashutosh-mishra) [inline gbpservice/nfp/orchestrator/modules/device_orchestrator.py]: Define it in a constant file.

File: gbpservice/nfp/orchestrator/modules/device_orchestrator.py [hunk 11]
```diff
@@ -578,17 +591,10 @@ def _prepare_device_data_from_nfp_context(self, nfp_context):
...
+            ports = self._make_ports_dict(consumer, provider, 'pt')
+        else:
+            ports = self._make_ports_dict(consumer, provider, 'port')
```
Comment (ashutosh-mishra) [inline gbpservice/nfp/orchestrator/modules/device_orchestrator.py]: Can we fill this info in node driver,
Its better to avoid such kind of checks in orchestrator.
Check with @njagadish once.

File: gbpservice/nfp/configurator/drivers/firewall/asav/asav_fw_driver.py (hunk 3 above)
Comment (ashutosh-mishra) [inline gbpservice/nfp/configurator/drivers/firewall/asav/asav_fw_driver.py]: Can we convey the same bug to visibility team, else they might expect this to be present on chain launch. We can file a bug for this.
If this is a serious bug, then why we are proceeding further, why not return error.
Are these changes only for fixing chain launch, will there be another PR for visibility.

File: gbpservice/nfp/orchestrator/drivers/orchestration_driver.py (hunk 9 above)
Comment (sdharav) [inline gbpservice/nfp/orchestrator/drivers/orchestration_driver.py:1002]: @ashutosh-mishra  while melding, this for loop got four spaces back and is leading to go out of the else block context so just moved in.

File: gbpservice/nfp/orchestrator/drivers/orchestration_driver.py (hunk 6 above)
Comment (sdharav) [inline gbpservice/nfp/orchestrator/drivers/orchestration_driver.py:645]: @mak-454  that we discussed last time will change that.

File: gbpservice/nfp/orchestrator/drivers/orchestration_driver.py (hunk 4 above)
Comment (sdharav) [inline gbpservice/nfp/orchestrator/drivers/orchestration_driver.py:642]: it gets the result key from the frame work.

File: gbpservice/nfp/orchestrator/modules/device_orchestrator.py (hunk 11 above)
Comment (mak-454) [inline gbpservice/nfp/orchestrator/modules/device_orchestrator.py]: its filled from node driver, this is more of a local variable.

Comment (mak-454) [inline gbpservice/nfp/orchestrator/modules/device_orchestrator.py]: here we always need gbp port - remove 'if', 'else'

**Comment by njagadish** [comment]
Changes are fine.
